### ISSUE TITLE
feature request: decouple `json` option from response body parsing

### ISSUE THREAD

**Issue by aoberoi** [sindresorhus/got#467]
currently, the `json` option both serializes the request body as content-type `application/json` *and* parses the response body assuming it is content-type `application/json`.

while the former is pretty straight-forward, the response parsing completely ignores the server's authoritative statement (the `content-type` response header) about what the body contains. this means that if the server specifies, for example, `text/html` for the response content type, the package with throw an error. i think we can do better.

my proposal: change the meaning of the `json` to only affect the body serialization. if it i set, only attempt to `JSON.parse()` the response body if the `Content-Type` header specifies `application/json`, otherwise return a plain string (don' attempt to parse any other content-type specifically). i believe this would be a backwards-compatible change since all cases that resulted in success earlier would continue to result in success. some cases that used to be an error would now be success, and the shape of the `result.body` would change, but since that code path would never have been reached before, this should be compatible with existing code.

i'm happy to submit a PR if this is something the maintainers are willing to consider.

**Comment by aoberoi** [comment]
related issues:

#446 - i believe this could result in a more elegant solution for OP since they wouldn't have to catch an error in order to get a hold of the response body.

#383, #318 - these would be solved

**Comment by aoberoi** [comment]
well well, it looks like this discussion was already had in #168. it seems to have concluded that a better solution is something you all would be interested in. is the above proposal something you all can get behind? if not, the `json: 'auto'` idea sounds interesting to me too.

cc @floatdrop @sindresorhus @SamVerschueren @Autarc

**Comment by sindresorhus** [comment]
I don't think this is a good idea. `window.fetch` behaves the same way and forces you to explicitly choose the format you want `await (await fetch()).json()`.

**Comment by aoberoi** [comment]
@sindresorhus i don't think `window.fetch` behaves the same way, but please correct me if i misunderstood. in your example, the `.json()` is called on an instance of `Response`, and only signals that the response body can be parsed and returned as JSON. the request body is completely independent, and the caller of `fetch()` is responsible for serializing it on their own.

my proposal boils down to specifying that the `json` option is a statement about how the caller wants to serialize the request body (helpful, since `fetch()` doesn't provide this). but if we want to be more aligned with `window.fetch`, then we shouldn't even attempt to parse the response body until the caller somehow signals that. all this proposal says is that we shouldn't fail when the caller doesn't signal and the server says the response shouldn't be parsed as JSON.

it would be helpful to know if you disagree with the specific proposal (and i'm happy to continue discussing) or if you disagree that there is a need. for the latter, i think the justification is both that this keeps coming up (see linked issues) and that `window.fetch` would allow me to do something that currently `got` does not.

**Comment by brandon93s** [comment]
The `json` option is a shorthand intended to support JSON request / JSON response APIs. Fortunately, headers, parsing and encoding can all be configured by the user when the API they are interacting with does not conform to JSON/JSON. 

When comparing to other libraries:

#### fetch
- Must serialize your own data
- Must read data as json with `.json()`

#### request
- `json` indicates write and read as JSON (similar to got)

#### axios
- supports a `responseType` (default is json)
- sets content type and serializes if `data` is an object

Since the option is intended to be a convenience and does not prohibit users from configuring got to handle more specific use-cases, I would prefer the API remain as-is.
